migration/idempotent/001-base.sql

```diff
@@ -3384,73 +3384,193 @@ $$
+
+    CREATE OR REPLACE VIEW _prom_catalog.metric_chunks_that_need_to_be_compressed AS
+    SELECT
+        m.id AS metric_id,
+        m.metric_name,
+        m.hypertable_id,
+        c.chunks_to_compress
+    FROM
+    (
+        SELECT
+            m.id,
+            m.metric_name,
+            m.table_name,
+            h.id AS hypertable_id
+        FROM _prom_catalog.metric m
+        INNER JOIN _timescaledb_catalog.hypertable h ON (h.schema_name OPERATOR(pg_catalog.=) 'prom_data' AND h.table_name OPERATOR(pg_catalog.=) m.table_name)
+        WHERE m.is_view IS FALSE
+        AND (m.delay_compression_until IS NULL OR m.delay_compression_until OPERATOR(pg_catalog.<) now())
+        AND h.compression_state OPERATOR(pg_catalog.=) 1 -- compression enabled
+        ORDER BY random() -- random metric order
+    ) m
+    INNER JOIN LATERAL -- lateral to force looking up chunks 1 metric at a time in nested loop
+    (
+        SELECT -- array of objects in order from oldest to newest
+            jsonb_agg
+            (
+                jsonb_build_object
+                (
+                    'chunk_id', x.chunk_id,
+                    'schema_name', x.schema_name,
+                    'table_name', x.table_name,
+                    'range_end', x.range_end,
+                    'chunk_num', x.chunk_num
+                )
+                ORDER BY x.chunk_num DESC -- oldest chunks first
+            )
+            FILTER
+            (
+                WHERE x.chunk_num OPERATOR(pg_catalog.>) 1 -- don't compress the most recent chunk
+                AND x.range_end OPERATOR(pg_catalog.<=) now() - interval '1 hour' -- don't compress chunks that aren't at least 1 hour old
+            )
+            AS chunks_to_compress
+        FROM
+        (
+            SELECT
+                c.id AS chunk_id,
+                c.schema_name,
+                c.table_name,
+                _timescaledb_internal.to_timestamp(ds.range_end) AS range_end,
+                row_number() OVER (ORDER BY ds.range_end desc) AS chunk_num -- number them from most recent to oldest i.e. newest will be 1
+            FROM _timescaledb_catalog.chunk c
+            INNER JOIN _timescaledb_catalog.chunk_constraint k ON (k.chunk_id OPERATOR(pg_catalog.=) c.id)
+            INNER JOIN _timescaledb_catalog.dimension_slice ds ON (ds.id OPERATOR(pg_catalog.=) k.dimension_slice_id)
+            WHERE c.hypertable_id OPERATOR(pg_catalog.=) m.hypertable_id
+            AND c.dropped IS FALSE
+            AND (c.status OPERATOR(pg_catalog.&) 1) OPERATOR(pg_catalog.!=) 1 -- not yet compressed
+        ) x
+    ) c ON (c.chunks_to_compress OPERATOR(pg_catalog.!=) '[]'::jsonb) -- don't bother with metrics that have no chunks to compress
+    ;
+    GRANT SELECT ON _prom_catalog.metric_chunks_that_need_to_be_compressed TO prom_maintenance;
+END;
+$block$;
+
+CREATE OR REPLACE FUNCTION _prom_catalog.compress_chunk(uncompressed_chunk regclass, if_not_compressed boolean DEFAULT false)
```

Comment:
ok. in that case, I'll just use `_prom_catalog.compress_chunk_for_hypertable`. I was hoping to avoid the extra query of the catalog, but it should not be a big deal.